### PR TITLE
SONARPY-616 FN on S5632: raise an issue when inferred type is 'str'

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**

### CODE REVIEW

File: python-frontend/src/main/java/org/sonar/plugins/python/api/symbols/ClassSymbol.java [hunk 1]
```diff
@@ -39,4 +39,7 @@
 
   @Beta
   boolean isOrExtends(ClassSymbol other);
+
+  @Beta
+  boolean canBeOrExtend(String fullyQualifiedClassName);
```
Comment (pynicolas) [inline python-frontend/src/main/java/org/sonar/plugins/python/api/symbols/ClassSymbol.java:44]: What about using this new method in other places? Are you going to open a new PR for that?

**Comment by pynicolas** [comment]
Good job @andrea-guarino-sonarsource! I think this PR is pushing our symbol model in the right direction.

File: python-checks/src/main/java/org/sonar/python/checks/IncorrectExceptionTypeCheck.java [hunk 2]
```diff
@@ -46,39 +46,35 @@ public void initialize(Context context) {
...
       }
-      if (raisedExpression.is(Tree.Kind.NAME)) {
-        Symbol symbol = ((Name) raisedExpression).symbol();
-        if (!inheritsFromBaseException(symbol, raisedExpression)) {
-          ctx.addIssue(raiseStatement, MESSAGE);
-        }
+      Symbol symbol = null;
+      if (raisedExpression instanceof HasSymbol) {
+        symbol = ((HasSymbol) raisedExpression).symbol();
+      } else if (raisedExpression.is(Tree.Kind.CALL_EXPR)) {
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/IncorrectExceptionTypeCheck.java:56]: Indeed, that's a mistake. I will remove it, thanks!

File: python-frontend/src/main/java/org/sonar/plugins/python/api/symbols/ClassSymbol.java (hunk 1 above)
Comment (andrea-guarino-sonarsource) [reply python-frontend/src/main/java/org/sonar/plugins/python/api/symbols/ClassSymbol.java:44]: Yes, good point. I think we have a few rules that can benefit from this API

File: python-checks/src/main/java/org/sonar/python/checks/IncorrectExceptionTypeCheck.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/IncorrectExceptionTypeCheck.java:56]: Looking at ruling results, I realized we may miss issues like the following:
```python
def f():
  raise NotImplemented("foo bar")
```
I'll keep the code like this and I'll update unit test to make it more explicit

File: python-frontend/src/main/java/org/sonar/python/semantic/SymbolUtils.java [hunk 3]
```diff
@@ -106,33 +107,66 @@ public static String fullyQualifiedModuleName(String packageName, String fileNam
...
+  private static boolean isBuiltinTypeshedFile(PythonFile pythonFile) {
+    return isTypeShedFile(pythonFile) && pythonFile.fileName().isEmpty();
+  }
+
+  private static boolean isTypingFile(PythonFile pythonFile) {
+    return isTypeShedFile(pythonFile) && pythonFile.fileName().equals("typing");
+  }
+
+  @CheckForNull
+  private static Symbol getParentSymbol(Argument argument) {
```
Comment (pynicolas) [inline python-frontend/src/main/java/org/sonar/python/semantic/SymbolUtils.java:157]: I don't understand what we mean by "parent" here.

Comment (andrea-guarino-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/SymbolUtils.java:157]: "parent" is refererring to "parent class (of the current class)". I understand the confusion, I'll rename it "getSymbolFromArgument"

Comment (andrea-guarino-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/SymbolUtils.java]: Fixed in 85f07023b31df792b45e866765cbcc6eb4e45c09

**Pull request merged**